src/protego.py

```diff
@@ -0,0 +1,406 @@
+            if len(parts) == 2:
+                rate, time_period = parts
+            else:
+                rate, time_period = parts[0], ''
+
+            requests, seconds = rate.split('/')
+            requests = int(requests)
+            time_unit = seconds[-1].lower()
+            seconds = int(seconds[:-1])
+            if time_unit == 'm':
+                seconds *= 60
+            elif time_unit == 'h':
+                seconds *= 3600
+            elif time_unit == 'd':
+                seconds *= 86400
+
+            start_time = None
+            end_time = None
+            if time_period:
+                start_time, end_time = time_period.split('-')
+                start_time = time(int(start_time[:2]), int(start_time[-2:]))
+                end_time = time(int(end_time[:2]), int(end_time[-2:]))
+        except Exception:
+            # Value is malformed, do nothing.
+            logger.debug("Malformed rule at line {} : cannot set request rate using '{}'. "
+                         "Ignoring this rule.".format(self._parser_instance._total_line_seen, value))
+            return
+
+        self._req_rate = RequestRate(requests, seconds, start_time, end_time)
+
+
+class Protego(object):
+
+    def __init__(self):
+        # A dict mapping user agents (specified in robots.txt) to rule sets.
+        self._user_agents = {}
+
+        # Preferred host specified in the robots.txt
+        self._host = None
+
+        # A list of sitemaps specified in the robots.txt
+        self._sitemap_list = []
+
+        # A memoization table mapping user agents (used in queries) to matched rule sets.
+        self._matched_rule_set = {}
+
+        self._total_line_seen = 0
+        self._invalid_directive_seen = 0
+        self._total_directive_seen = 0
+
+    @classmethod
+    def parse(cls, content):
+        o = cls()
+        o._parse_robotstxt(content)
+        return o
+
+    def _parse_robotstxt(self, content):
+        lines = content.splitlines()
+
+        # A list containing all the user agents of the current record group.
+        current_user_agents = []
+
+        # Last encountered rule irrespective of whether it was valid or not.
+        previous_rule_field = None
+
+        for line in lines:
+            self._total_line_seen += 1
+
+            # Remove the comment portion of the line
+            hash_pos = line.find('#')
+            if hash_pos != -1:
+                line = line[0: hash_pos].strip()
+
+            # Whitespace at the beginning and at the end of the line is ignored.
+            line = line.strip()
+            if not line:
+                continue
+
+            # Format for a valid robots.txt rule is "<field>:<value>"
+            if line.find(':') != -1:
+                field, value = line.split(':', 1)
+            else:
+                # We will be generous here and give it a second chance.
+                parts = line.rsplit(' ', 1)
+                if not len(parts) == 2:
```

Comment:
Shouldn’t we allow also cases where the number of parts is higher than 2? For example: `User-agent Mozilla Firefox`